analyzer/src/steps/resolve.rs

```diff
@@ -287,53 +340,219 @@ impl<'a, 'e> SymbolResolver<'a, 'e> {
                 .engine
                 .get_environment(origin)
                 .expect("Environment declared an unknown parent");
-            let name = origin_env
+
+            let symbol_name = origin_env
                 .variables
-                .get_symbol_name(object_id)
+                .get_external_symbol_name(object_id)
                 .expect("Unknown object name");
 
             // Go up the parent chain until we find the symbol or we reach the root
-            let mut current = origin_env;
-            while let Some((module, env)) = current
-                .parent
-                .and_then(|id| self.engine.get_environment(id).map(|env| (id, env)))
-            {
-                // Locals symbols are always treated first, before imports.
-                // The current environment might already owns the resolution result as a global symbol.
-                // This happens only if it used it, so we ignore that fact here to always solve external
-                // symbols via imports.
-                if !env.has_strict_declaration_order() {
-                    if let Some(Symbol::Local(local)) = env.variables.get_exported(name) {
-                        object.resolved = Some(ResolvedSymbol {
-                            module,
-                            object_id: local,
-                        });
-                        continue 'symbol;
+            let mut current = Some((origin, origin_env));
+            while let Some((env_id, env)) = current {
+                let mut result = SymbolResolutionResult::NotFound;
+
+                if env_id != origin {
+                    // Locals symbols are always treated first, before imports.
+                    // The current environment might already owns the resolution result as a global symbol.
+                    // This happens only if it used it, so we ignore that fact here to always solve external
+                    // symbols via imports.
+                    //
+                    // We omit this resolution for origin environments as the resolution of their own locals
+                    // is done by the collection phase
+                    result = Self::resolve_symbol_from_locals(env_id, env, symbol_name);
+                }
+
+                if result == SymbolResolutionResult::NotFound {
+                    if let Some(imports) = &self.relations.get_imports(&env_id) {
```

Comment:
```suggestion
                    if let Some(imports) = self.relations.imports.get(&env_id) {
```
Requires to exposes `imports` (or move it outside as mentioned above)